File: prisoners.java
import java.util.*;

class Prisoners {
	static int MAX = 1024;
	static Map<String, Integer>mem = new HashMap<>();

	static int solve(int a, int b, int [] r) {
		String key = a + " " + b;
		if (mem.containsKey(key)) {
			return mem.get(key);
		}
		int result = Integer.MAX_VALUE;
		for (int i = 0; i < r.length; i++) {
			if (r[i] >= a && r[i] <= b) {
				int c1 = solve(a, r[i] - 1, r);
				int c2 = solve(r[i] + 1, b, r);
				result = Math.min(result, b-a + c1 + c2);
			}
		}
		if (result == Integer.MAX_VALUE)
			result = 0;
		mem.put(key, result);
		return result;
	}

	public static void main(String [] args) {
		Scanner scanner = new Scanner(System.in);
		int T = scanner.nextInt();
		for (int t = 1; t <= T; t++) {
			int P = scanner.nextInt();
			int N = scanner.nextInt();
			int a[] = new int[MAX];
			for (int i = 0; i < N; i++) {
				a[i] = scanner.nextInt();
			}
			mem.clear();
			int result = solve(1, P, a);
			System.out.println("Case #" + t + ": " + result);
		}
	}

}

File: prisoners.cs
using System;
using System.Collections;
using System.Collections.Generic;

class MainClass {

	static void Main() {
		int [] p = {3, 5};
		int result = Solve(p, 1, 9);
		Console.Write(result);
	}

	static int Solve(int [] p, int L, int R) {
		if (L > R) return 0;
		string key = "" + L + '-' + R;
		if (cache.ContainsKey(key)) return cache[key];
		int min = 0;
		for (int i = 0; i < p.Length; i++) {
			if (p[i] >= L && p[i] <= R) {
				int resultL = Solve(p, L, p[i]-1);
				int resultR = Solve(p, p[i]+1, R);
				int cost = resultL + resultR + (R-L);
				if (min == 0) min = cost;
				if (min > cost) min = cost;
			}
		}
		cache[key] = min;
		return min;
	}

	static Dictionary<string, int> cache = new Dictionary<string, int>();
}

